okio/src/main/java/okio/BufferCursor.java

```diff
@@ -0,0 +1,156 @@
+  public long size = -1L;
+
+  private Segment segment;
+  public long offset = -1L;
+  public byte[] data;
+  public int pos = -1;
+  public int limit = -1;
+
+  public void acquireReadOnly(Buffer buffer) {
+    if (this.buffer != null) throw new IllegalStateException("already acquired");
+
+    this.buffer = buffer;
+    this.size = buffer.size;
+  }
+
+  /**
+   * Seeks to the next range of bytes, advancing the offset by {@code limit - pos}. Returns the size
+   * of the readable range (at least 1), or -1 if we have reached the end of the buffer and there
+   * are no more bytes to read.
+   */
+  public int next() {
+    if (offset == size) throw new IllegalStateException();
+    if (offset == -1L) return seek(0L);
+    return seek(offset + (limit - pos));
+  }
+
+  /**
+   * Reposition the cursor so that the data at {@code offset} is readable at {@code data[pos]}.
+   * Returns the number of bytes readable in {@code data} (at least 1), or -1 if we have reached the
+   * end of the buffer and there are no more bytes to read.
+   */
+  public int seek(long offset) {
+    if (offset < 0 || offset > size) {
+      throw new ArrayIndexOutOfBoundsException(String.format("offset=%s > size=%s", offset, size));
+    }
+
+    if (offset == size) {
+      this.segment = null;
+      this.offset = offset;
+      this.data = null;
+      this.pos = -1;
+      this.limit = -1;
+      return -1;
+    }
+
+    // Navigate to the segment that contains `offset` starting from our current segment if possible.
+    Segment next;
+    long nextOffset;
+    if (segment == null || this.offset > offset) {
+      next = buffer.head;
+      nextOffset = 0L;
+    } else {
+      next = this.segment;
+      nextOffset = this.offset - (pos - next.pos);
+    }
+    while (offset >= nextOffset + next.limit - next.pos) {
+      nextOffset += (next.limit - next.pos);
+      next = next.next;
+    }
+
+    // Update this cursor to the requested offset within the found segment.
+    // TODO(jwilson): get a segment that is OWNED and NOT SHARED if we're writing.
+    this.segment = next;
+    this.offset = offset;
+    this.data = next.data;
+    this.pos = next.pos + (int) (offset - nextOffset);
+    this.limit = next.limit;
+    return limit - pos;
+  }
+
+  /**
+   * Add additional capacity to the end of this buffer until its new size is {@code newSize}. This
+   * must be at least the current size. Newly added capacity may span multiple segments.
+   *
+   * <p>Warning: it its the caller’s responsibility to write new data to every byte of the
+   * newly-allocated capacity. Failure to do so may cause serious security problems as the data
+   * in the returned buffers is not zero filled. The buffers may contain dirty pooled segments that
+   * hold very sensitive data from other parts of the current process.
+   *
+   * TODO(jwilson): just zero-fill this to be safe? Or expose a parameter to make that easy?
+   */
+  public void enlarge(long newSize) {
```

Comment:
new! And super dangerous!

Reply:
In cases when you don't know how many bytes are going to be needed, there might need to be a way trim the buffer.  I'm thinking specifically of the `InflaterSource` and the `DeflaterSink` and how they might use a cursor.
```java
long offset = sink.size(); 
cursor.enlarge(...);
cursor.seek(offset);
int bytesInflated = inflater.inflate(cursor.data, cursor.pos, cursor.limit - cursor.pos);
cursor.trim(offset + bytesInflated);
```

Also, might be handy if enlarge returns the old size for things like `cursor.seek(cursor.enlarge(...))` which will get you immediately to the write point.

Reply:
Yeah, I agree that this needs to be `resize()` and should either grow or truncate the buffer.

Reply:
I also found this method helpful.  The idea being that I want the next available set of bytes for writing but don't want to allocate another segment unless needed.
```java
public void grow() {
  Segment tail = buffer.writableSegment(1);
  resize(size + Segment.SIZE - tail.limit);
}
```